prov/gni/src/gnix_av.c

```diff
@@ -36,13 +36,449 @@
 #include <string.h>
+#include <inttypes.h>
 
 #include "gnix.h"
 #include "gnix_util.h"
 
+
+/*******************************************************************************
+ * Forward declarations of ops structures.
+ ******************************************************************************/
+static struct fi_ops_av gnix_av_ops;
+static struct fi_ops gnix_fi_av_ops;
+
+/*******************************************************************************
+ * Helper functions.
+ ******************************************************************************/
+/*
+ * TODO: Check for named AV creation and check RX CTX bits.
+ */
+static int gnix_verify_av_attr(struct fi_av_attr *attr)
+{
+	int ret = FI_SUCCESS;
+
+	switch (attr->type) {
+	case FI_AV_TABLE:
+	case FI_AV_MAP:
+		break;
+	default:
+		ret = -FI_EINVAL;
+		break;
+	}
+
+	return ret;
+}
+
+/*
+ * Check the capacity of the internal table used to represent FI_AV_TABLE type
+ * address vectors. Initially the table starts with a capacity and count of 0
+ * and the capacity increases by roughly double each time.
+ */
+static int gnix_check_capacity(struct gnix_fid_av *av, size_t count)
+{
+	struct addr_entry *addrs;
+	size_t capacity = av->capacity;
+
+	/*
+	 * av->count + count is the amount of used indices after adding the
+	 * count items.
+	 */
+	while (capacity < av->count + count) {
+		/*
+		 * Handle initial capacity of 0, by adding 1.
+		 */
+		capacity = capacity * 2 + 1;
+	}
+
+	/*
+	 * Don't need to grow the table.
+	 */
+	if (capacity == av->capacity) {
+		return FI_SUCCESS;
+	}
+
+	addrs = realloc(av->table, capacity * sizeof(*addrs));
+	if (!addrs) {
+		return -FI_ENOMEM;
+	}
+
+	/*
+	 * Update table and capacity to reflect new values.
+	 */
+	av->table = addrs;
+	av->capacity = capacity;
+
+	return FI_SUCCESS;
+}
+
+/*******************************************************************************
+ * FI_AV_TABLE specific implementations.
+ ******************************************************************************/
+static int table_insert(struct gnix_fid_av *int_av, const void *addr,
+			size_t count, fi_addr_t *fi_addr, uint64_t flags,
+			void *context)
+{
+	struct gnix_ep_name *temp;
+	int ret = count;
+	size_t index;
+	size_t i;
+
+	if (gnix_check_capacity(int_av, count)) {
+		ret = -FI_ENOMEM;
+		goto err;
+	}
+
+	for (index = int_av->count, i = 0; i < count; index++, i++) {
+		temp = &((struct gnix_ep_name *)addr)[i];
+		int_av->table[index].addr = &temp->gnix_addr;
+		int_av->table[index].valid = true;
+		fi_addr[i] = index;
+	}
+
+	int_av->count += count;
+
+err:
+	return ret;
+}
+
+/*
+ * TODO: Actually free memory.
+ * Currently only marks as 'not valid'. Should actually free memory.
+ */
+static int table_remove(struct gnix_fid_av *int_av, fi_addr_t *fi_addr,
+			size_t count, uint64_t flags)
+{
+	size_t index;
+	size_t i;
+
+	for (i = 0; i < count; i++) {
+		index = (size_t) fi_addr[i];
+		if (index < int_av->count) {
+			int_av->table[index].valid = false;
```

Comment:
The case for fi_av_remove is kind of interesting because according to the man page it is only supposed to return 0 (FI_SUCCESS) on success or a negative error code if something went wrong. Since fi_av_remove takes a list of items, if any of those error then the error code would correspond for a single removal. Not sure exactly how to handle this. It would make more sense if the return value was the number of items successfully removed.